### PR TITLE
BUG: Fix weighted rolling variance implementation

### LINKED ISSUE #53273: BUG:
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np

df = pd.Series(np.random.normal(size=1200))
df_std_stock = (
    df.rolling(window=220, min_periods=100, win_type='exponential')
      .std(tau=-(25 / np.log(2)), center=0, sym=False)
)
```

### Issue Description

Hey, Team! The series of standard deviations produced by the rolling calculation with custom weights does not look right and I hope that a visual demonstration illustrates that best. 
image

The blue line is the result of the calculation above and it is visible how after `window` observations the series no longer produces sensible results. Odd and strong changes in behavior are visible every pass over a window multiple, which is even more suspicious. The orange series is what I would expect the result to be and the calculation snippet that produced it is in the box below. The algorithm for the expected behavior is known to be problematic though.

Thank you for taking the time to consider my report and please let me know if further details are required. Lover your package and really hope that his helps making it even better.

### Expected Behavior

```python
import pandas as pd
import numpy as np

np.random.seed(42)
df = pd.Series(np.random.normal(size=1200))
df_std_custom = (
    (df ** 2).rolling(window=220, min_periods=100, win_type='exponential')
             .mean(tau=-(25 / np.log(2)), center=0, sym=False)
)
df_std_custom_mean = (
    df.rolling(window=220, min_periods=100, win_type='exponential')
      .mean(tau=-(25 / np.log(2)), center=0, sym=False)
)
df_std_custom = (df_std_custom - df_std_custom_mean**2).pow(0.5)
```

### Installed Versions

INSTALLED VERSIONS
------------------
commit           : 37ea63d540fd27274cad6585082c91b1283f963d
python           : 3.9.16.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19044
machine          : AMD64
processor        : Intel64 Family 6 Model 141 Stepping 1, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : English_United States.1252

pandas           : 2.0.1
numpy            : 1.22.4
pytz             : 2023.3
dateutil         : 2.8.2
setuptools       : 66.0.0
pip              : 23.0.1
Cython           : 0.29.34
pytest           : 7.3.1
hypothesis       : 6.75.3
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.13.2
pandas_datareader: None
bs4              : 4.12.2
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.7.1
numba            : None
numexpr          : 2.8.4
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 12.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.7.1
snappy           : None
sqlalchemy       : None
tables           : 3.8.0
tabulate         : None
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : 2023.3
qtpy             : None
pyqt5            : None

### LINKED ISSUE #54333: BUG:
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import numpy as np
import pandas as pd
from scipy import signal

def equal(win_len: int, **kwargs):
    assert win_len == 5
    return np.array([1.0, 1.0, 1.0, 1.0, 1.0])

def linear(win_len: int, **kwargs):
    assert win_len == 5
    return np.array([0.2, 0.4, 0.6, 0.8, 1.0])

signal.windows.equal_weight = equal
signal.windows.linear_decay = linear

s = pd.Series([1.0, 0.0] * 10)

print(list(s))
print(list(s.rolling(window=5).var(ddof=0)))
print(list(s.rolling(window=5, win_type="equal_weight").var(ddof=0)))
print(list(s.rolling(window=5, win_type="linear_decay").var(ddof=0)))
```

### Issue Description

The 4 prints of the code snippet generates the following
```python
[1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0]
[nan, nan, nan, nan, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24]
[nan, nan, nan, nan, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24]
[nan, nan, nan, nan, 0.24000000000000005, 0.24888888888888897, 0.2222222222222223, 0.24888888888888902, 0.1955555555555556, 0.24000000000000007, 0.248888888888889, 0.22222222222222232, 0.24888888888888897, 0.19555555555555562, 0.23999999999999996, 0.24888888888888888, 0.2222222222222222, 0.24888888888888897, 0.1955555555555555, 0.23999999999999996]
```

However the last line is wrong. The alternating 1/0 sequence, when in a window of size 5 (i.e. either `10101` or `01010`), should have a variance of 0.24 when the weight is either all equal or linear decay (feel free to verify this). This means the last 3 prints should all generate
```python
[nan, nan, nan, nan, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24]
```
but it's not the case as the elments in the last line are pretty off 0.24 - the first non-Nan value is close enough, but the four numbers after that are pretty off.

### Expected Behavior

```python
list(s.rolling(window=5, win_type="linear_decay").var(ddof=0))
```
should be
```python
[nan, nan, nan, nan, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24]
```
or something _really_ close to that

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 0f437949513225922d851e9581723d82120684a6
python           : 3.10.12.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.15.0-1042-azure
Version          : #49-Ubuntu SMP Tue Jul 11 17:28:46 UTC 2023
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 2.0.3
numpy            : 1.21.5
pytz             : 2022.1
dateutil         : 2.8.2
setuptools       : 59.6.0
pip              : 22.0.2
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.8.0
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 7.31.1
pandas_datareader: None
bs4              : 4.10.0
bottleneck       : None
brotli           : 1.0.9
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.5.1
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.8.0
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : 2023.3
qtpy             : None
pyqt5            : None

</details>

Note the issue is also there if you use the latest version of `numpy` and `scipy` (which are `1.25.2` and `1.11.1` respectively as of filing)

### PR DESCRIPTION
**Pull request by ihsansecer**
- [x] closes #53273, closes #54333
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.
---

Outputs from the code snippets in the given issues are below

#### Code
```python
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

np.random.seed(42)
df = pd.Series(np.random.normal(size=1200))
df_std_stock = (
    df.rolling(window=220, min_periods=100, win_type='exponential')
    .std(tau=-(25 / np.log(2)), center=0, sym=False)
)

df_std_custom = (
    (df ** 2).rolling(window=220, min_periods=100, win_type='exponential')
    .mean(tau=-(25 / np.log(2)), center=0, sym=False)
)
df_std_custom_mean = (
    df.rolling(window=220, min_periods=100, win_type='exponential')
    .mean(tau=-(25 / np.log(2)), center=0, sym=False)
)
df_std_custom = (df_std_custom - df_std_custom_mean ** 2).pow(0.5)

plt.plot(df_std_stock.index, df_std_stock.values, label="df_std_stock", linewidth=3)
plt.plot(df_std_custom.index, df_std_custom.values, label="df_std_custom", linestyle='--')
plt.legend()
plt.show()

```
#### Output
test

#### Code
```python
import numpy as np
import pandas as pd
from scipy import signal

def equal(win_len: int, **kwargs):
    assert win_len == 5
    return np.array([1.0, 1.0, 1.0, 1.0, 1.0])

def linear(win_len: int, **kwargs):
    assert win_len == 5
    return np.array([0.2, 0.4, 0.6, 0.8, 1.0])

signal.windows.equal_weight = equal
signal.windows.linear_decay = linear

s = pd.Series([1.0, 0.0] * 10)

print(list(s))
print(list(s.rolling(window=5).var(ddof=0)))
print(list(s.rolling(window=5, win_type="equal_weight").var(ddof=0)))
print(list(s.rolling(window=5, win_type="linear_decay").var(ddof=0)))
```
#### Output
```
[1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0]
[nan, nan, nan, nan, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24]
[nan, nan, nan, nan, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24, 0.24]
[nan, nan, nan, nan, 0.24000000000000005, 0.24000000000000005, 0.24000000000000005, 0.24000000000000005, 0.24000000000000005, 0.24000000000000005, 0.24000000000000005, 0.24000000000000005, 0.24000000000000005, 0.24000000000000005, 0.24000000000000005, 0.24000000000000005, 0.24000000000000005, 0.24000000000000005, 0.24000000000000005, 0.24000000000000005]
```

### CODE REVIEW

**Comment by mroeschke** [comment]
Can you show if there is any performance difference before and after this change?

File: pandas/_libs/window/aggregations.pyx [hunk 1]
```diff
@@ -1690,44 +1642,38 @@ def roll_weighted_var(const float64_t[:] values, const float64_t[:] weights,
...
-        float64_t[:] output
+        float64_t val, w
+        Py_ssize_t in_i, win_i, add_i, n, win_n
+        float64_t[:] output, t, mean, sum_w, nobs
 
     n = len(values)
     win_n = len(weights)
+
     output = np.empty(n, dtype=np.float64)
+    t = np.zeros(n, dtype=np.float64)
```
[the remark this answers is not in the document]
Comment (ihsansecer) [reply pandas/_libs/window/aggregations.pyx:1653]: Ah yeah it should be possible I think

**Comment by ihsansecer** [comment]
> Can you show if there is any performance difference before and after this change?

| Change   | Before [ebca6dfb] <main>   | After [a695dc48] <fix-rolling-std-custom-weights>   |   Ratio | Benchmark (Parameter)                           |
|----------|----------------------------|-----------------------------------------------------|---------|-------------------------------------------------|
| +        | 2.18±0.02ms                | 764±4ms                                             |  351.17 | rolling.WeightedVar.time_method(1000, 'boxcar') |
| +        | 2.25±0.02ms                | 768±4ms                                             |  341.83 | rolling.WeightedVar.time_method(1000, 'triang') |
| +        | 2.24±0.03ms                | 73.6±0.3ms                                          |   32.83 | rolling.WeightedVar.time_method(100, 'boxcar')  |
| +        | 2.36±0.3ms                 | 74.6±0.1ms                                          |   31.57 | rolling.WeightedVar.time_method(100, 'triang')  |
| +        | 2.96±0.9ms                 | 5.09±0.06ms                                         |    1.72 | rolling.WeightedVar.time_method(10, 'boxcar')   |
| +        | 3.07±0.9ms                 | 5.07±0.04ms                                         |    1.65 | rolling.WeightedVar.time_method(10, 'triang')   |

---

```python
class WeightedVar:
    params = (
        [10, 100, 1000],
        ["triang", "boxcar"]
    )
    param_names = ["window", "win_type"]

    def setup(self, window, win_type):
        N = 10**5
        arr = np.random.random(N)
        obj = pd.Series(arr)
        self.window = obj.rolling(window, win_type=win_type)

    def time_method(self, window, win_type):
        self.window.var()

File: pandas/_libs/window/aggregations.pyx (hunk 1 above)
[the remark this answers is not in the document]
Comment (ihsansecer) [reply pandas/_libs/window/aggregations.pyx]: This is done

**Comment by jreback** [comment]
the new implementation is O^2 
this is not good at all

**Comment by ihsansecer** [comment]
> the new implementation is O^2 this is not good at all

@jreback I don't think there is a better solution. Let me explain what is wrong with the previous implementation and what this implementation does using an example.

Let's say we have:
```python
values = [0, 1, 2, 3]
weights = [10, 20]
```
We are supposed to update the sum of weights, mean and t each time a value and a weight is added (or removed) to the window as per the [paper](https://dl.acm.org/doi/pdf/10.1145/359146.359153). After each pair is added (or removed) to the window we can calculate weighted variance.

The previous implementation follows the steps below:
```python
var, w = 0, 10
add(var, w)
output[0] = calculate()

var, w = 1, 20
add(var, w)
output[1] = calculate()

var, w = 2, 10
add(var, w)
var_pre, w_pre = 0, 10
remove(var_pre, w_pre)
output[2] = calculate()

var, w = 3, 20
add(var, w)
var_pre, w_pre = 1, 20
remove(var_pre, w_pre)
output[3] = calculate()
```
Single pass is possible if adding the head pair (or removing the tail) to the window is enough but it is not. Since all of the pairings are different, we need to first remove all the pairs that are previously added (or start from scratch) and then add the pairs we want.


What the new implementation does is:
```python
var, w = 0, 20
add(var, w)
output[0] = calculate()

reset_t_sumw_mean()
var, w = 0, 10
add(var, w)

var, w = 1, 20
add(var, w)
output[1] = calculate()

reset_t_sumw_mean()
var, w = 1, 10
add(var, w)

var, w = 2, 20
add(var, w)
output[2] = calculate()

reset_t_sumw_mean()
var, w = 2, 10
add(var, w)

var, w = 3, 20
add(var, w)
output[3] = calculate()
```

**Comment by ihsansecer** [comment]
@jreback @WillAyd @mroeschke any other comments/concerns?

**Comment by mroeschke** [comment]
Is there a way to fix this while keeping the current algorithm? The slowdown is concerning for more common cases where this was already correct

**Comment by ihsansecer** [comment]
> Is there a way to fix this while keeping the current algorithm?

I don't think there is. I tried to explain what was wrong with it (and why it is faster) in the previous comment. I think it should be incorrect almost all the time because it is calculating variance for the wrong pairs of weights and values

**Comment by WillAyd** [comment]
From the paper linked:

> It is frequently convenient to calculate these quanti-
ties by a method that, unlike direct implementation of
the above definitions, requires only one pass through the
set of data pairs (X~, W~). Such a method is also useful
where it is necessary tc, update (rather than recalculate
ab initio) the mean and variance after the data set has
been extended to include another data pair.

But isn't this algorithm requiring multiple passes instead of simply updating records as the window expands? The scaling and performance at 1000 rows makes this pretty impractical to use for usual datasets I think

**Comment by ihsansecer** [comment]
> The scaling and performance at 1000 rows makes this pretty impractical to use for usual datasets I think

@WillAyd It is not the performance at 1000 rows. The number of rows is constant at 100,000 and the varying one is the length of weights. 

> But isn't this algorithm requiring multiple passes instead of simply updating records as the window expands?

It still calculates variance in one pass (over the same value and weight pairs). The problem is nothing here is expanding really. We **have to** recalculate variance for each window because pairs are different in each window. I have an excalidraw board below showing how this works using the same example. Hope this makes sense.

Untitled-2023-09-30-1837


By the way, both weighted moving variance and weighted sum have the same complexity O(wn) where w is the number of weights and n is the number of rows.
https://github.com/pandas-dev/pandas/blob/6f0cd8d9ed251aa80f7415c565176fc33487110a/pandas/_libs/window/aggregations.pyx#L1495-L1500

**Comment by mroeschke** [comment]
Thanks for the PR here but it has appeared to gone stale. I think the discussion on the PR is better suited for the original issue(s) so let's move the discussion there before proceeding with this

**Pull request closed without merging**